types/src/validator_verifier.rs

```diff
@@ -199,47 +207,72 @@ impl ValidatorVerifier {
                 .get(addr)
                 .ok_or(VerifyError::UnknownAuthor)?;
             masks.set(index as u16);
-            pub_keys.push(self.validator_infos[index].public_key());
+            pub_keys.insert(index, self.validator_infos[index].public_key());
             sigs.push(sig.clone());
         }
         // Perform an optimistic aggregation of the signatures without verification.
         let aggregated_sig = bls12381::Signature::aggregate(sigs)
             .map_err(|_| VerifyError::FailedToAggregateSignature)?;
 
-        let aggregated_key =
-            PublicKey::aggregate(pub_keys).map_err(|_| VerifyError::FailedToAggregatePubKey)?;
+        let ordered_pub_keys: Vec<_> = pub_keys.into_values().collect_vec();
+
+        let aggregated_key = PublicKey::aggregate(ordered_pub_keys.clone())
+            .map_err(|_| VerifyError::FailedToAggregatePubKey)?;
         Ok((
-            MultiSignature::new(masks, Some(aggregated_sig)),
+            AggregatedSignature::new(masks, Some(aggregated_sig)),
             aggregated_key,
+            ordered_pub_keys,
```

Comment:
Why do pubkeys need to be returned in an ordered fashion? Is it because messages are similarly-ordered and it makes it easier to verify the aggregate signature on those messages later on?

Reply:
>I don't think we should return the public key in this function, I'd prefer to call self.verify_ directly that goes through the normal verification flow

I am not sure I understand but this is the aggregation path and we are retuning the pub_keys so that we can  verify the aggregated sig. 

> Why do pubkeys need to be returned in an ordered fashion? Is it because messages are similarly-ordered and it makes it easier to verify the aggregate signature on those messages later on?

Its not about the ease, this is needed for the `verify_aggregate` function call.  The order of the message and the pub_keys needs to be the same.

Reply:
we don't need pubkey for verification, it should just call `self.verify_aggregated_signatures(messages, &aggregate_sig)`

Reply:
Yes, you are right but that means doing double verification for things like number of votes,  voting power etc, also we need to recreate the pub_keys, which are already available here. In addition, it semantically seems confusing to me to do a full verification in the aggregation path because the aggregation path already has done some partial verification. In fact, we don't need to verify anything at all after this - I have added this verification just as a good to have.

Reply:
the API is confusing to return the keys. calling the full verification can also make sure it works on the receiver side. we'll need the verification path after we switch to optimistic aggregation.

Reply:
Yes, we need to call the full verification once we switch to optimistic aggregation but that is unnecessary overhead for now. I have modified the API name and added relevant comments to make it clearer hopefully. LMK what you think.

Reply:
it's not about the naming, as the caller I find its confusing that I need to get the keys for aggregation. maybe we can just skip verify and minimize the api, we'll add the verification when we do optimistic aggregation.

Reply:
Aight, let's remove the verification - I was tempted to do that in the first place but kept it here just in case.